examples/generative/Density estimation using RealNVP.py

```diff
@@ -0,0 +1,251 @@
+        log_det_for = tf.zeros(y.shape[0])
+        x = y
+        for i in range(self.num_cl):
+            x_masked = x * self.masks[i]
+            reversed_mask = 1 - self.masks[i]
+            s, t = self.layers_list[i](x_masked)
+            s *= reversed_mask
+            t *= reversed_mask
+            x = x_masked + reversed_mask * (x * tf.exp(s) + t)
+            log_det_for += tf.reduce_sum(s, [1])
+
+        return x, log_det_for
+
+    # custom function defining the forward operation
+    def inverse(self, x):
+        # log determinant of the forward pass
+        log_det_inv = tf.zeros(x.shape[0])
+        y = x
+        for i in reversed(range(self.num_cl)):
+            y_masked = y * self.masks[i]
+            reversed_mask = 1 - self.masks[i]
+            s, t = self.layers_list[i](y_masked)
+            s *= reversed_mask
+            t *= reversed_mask
+            y = reversed_mask * (y - t) * tf.exp(-s) + y_masked
+            log_det_inv -= tf.reduce_sum(s, [1])
+
+        return y, log_det_inv
+
+    # log likelihood of the normal distribution + the log determinant of the jacobian
+    def log_likelihood(self, x):
+        y, logdet = self.inverse(x)
+        return self.distr.log_prob(y) + logdet
+
+    def log_loss(self, x):
+        return -tf.reduce_mean(self.log_likelihood(x))
+
+    def train_step(self, data):
+
+        with tf.GradientTape() as tape:
+            loss = self.log_loss(data)
+
+        # updating the variables
+        g = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(g, self.trainable_variables))
+        loss_tracker.update_state(loss)
+
+        return {"loss": loss_tracker.result()}
+
+
+"""
+## Initialization
+"""
+
+# initializing the masks, the layers and the prior distribution for the real nvp
+num_cl = 6
+masks = np.array([[0, 1], [1, 0]] * (num_cl // 2), dtype="float32")
+
+layers_list = [Coupling(2) for i in range(num_cl)]
+
+tfd = tfp.distributions
+distr = tfd.MultivariateNormalDiag(loc=[0.0, 0.0], scale_diag=[1.0, 1.0])
+
+"""
+## Model training
+"""
+loss_tracker = keras.metrics.Mean(name="loss")
+
+# Adam optimizer
+optimizer = tf.keras.optimizers.Adam(learning_rate=0.0001)
+
+model = Realnvp(layers_list, masks, distr)
+
+model.compile(optimizer=optimizer)
+
+
+class Loss_every_100_epochs(keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs={}):
+        if epoch % 500 == 0:
+            print(epoch, "loss", round(logs.get("loss"), 4))
+
+
+# Training model
+history = model.fit(
+    data,
+    batch_size=len(data),
+    epochs=7500,
+    callbacks=[Loss_every_100_epochs()],
+    verbose=0,
+)
+"""
+## Performance evaluation
+"""
+
+plt.figure(figsize=(15, 10))
+plt.plot(history.history["loss"])
+plt.title("model loss")
+plt.ylabel("loss")
+plt.xlabel("epoch")
+
+# from data to latent space
+z, _ = model.inverse(data)
+
+# from latent space to data
+samples = distr.sample(3000)
+x, _ = model.forward(samples)
```

Comment:
Use `model.predict(samples)` (which will work once you put your logic in `call`)